src/convert_hl7v2_fhir/controllers/er7/er7_extractor.py

```diff
@@ -0,0 +1,115 @@
+from datetime import datetime, timezone
+from typing import List
+import hl7apy
+from hl7apy.core import Message
+
+from convert_hl7v2_fhir.controllers.er7.exceptions import (
+    MissingNHSNumberError,
+    InvalidNHSNumberError,
+    MissingPointOfCareError,
+    MissingFacilityError,
+    MissingPatientClassError,
+    MissingAdmissionTypeError,
+    MissingTimeOfAdmissionError,
+    MissingFamilyNameError,
+    MissingGivenNameError,
+)
+from convert_hl7v2_fhir.controllers.utils import is_nhs_number_valid
+
+
+class ER7Extractor:
+    def __init__(self, er7_message: Message):
+        self.er7_message = er7_message
+
+    def nhs_number(self) -> str:
+        """There are two potential fields where NHS number can be:
+        1. PID.2 - Patient Id
+        2. PID.3 - Patient Identifier List"""
+
+        nhs_numbers = [
+            composite_id.id.value
+            for composite_id in self.er7_message.pid.patient_id_internal_id
+            if "NHSNMBR" in composite_id.value
+        ]
+        if "NHSNMBR" in self.er7_message.pid.patient_id_external_id.value:
+            nhs_numbers.append(self.er7_message.pid.patient_id_external_id.id.value)
+
+        if not nhs_numbers:
+            raise MissingNHSNumberError
+
+        try:
+            return next(
+                (
+                    nhs_number
+                    for nhs_number in nhs_numbers
+                    if is_nhs_number_valid(nhs_number)
+                )
+            )
+        except StopIteration:
+            raise InvalidNHSNumberError
```

Comment:
This feels quite wordy.  I don't think we'd lose anything by saying this:
```suggestion
        for nhs_number in nhs_numbers:
            if is_nhs_number_valid(nhs_number):
                return nhs_number
        raise InvalidNHSNumberError
```

If we wanted to go all functional we could do:
```suggestion
        try:
            return next(filter(is_nhs_number_valid, nhs_numbers))
        except StopIteration:
            raise InvalidNHSNumberError
```
But I'm not sure that gains us much, and introduces exceptions for logic which I'm not that keen on (despite Python disagreeing with me).